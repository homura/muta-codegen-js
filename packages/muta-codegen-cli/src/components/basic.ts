import { codegen as graphqlCodegen } from '@graphql-codegen/core';
import { CodegenPlugin, Types } from '@graphql-codegen/plugin-helpers/types';
import { buildSchema, parse, printSchema } from 'graphql';

interface Options {
  schemaSource: string;
  pluginsMap: {
    [name: string]: CodegenPlugin;
  };
  plugins: Types.ConfiguredPlugin[];
}

export function codegen(options: Options): Promise<string> {
  if (!options.schemaSource) return Promise.resolve('');

  return graphqlCodegen({
    schema: parse(printSchema(buildSchema(options.schemaSource))),
    plugins: options.plugins,
    pluginMap: options.pluginsMap,
    filename: '',
    documents: [],
    config: {
      scalars: {
        Address: 'Address',
        u64: 'u64',
        Hash: 'Hash',
        Uint64: 'Uint64',
        Null: "''",
        Hex: 'string',
        Uint32: 'string',
      },
      skipTypename: true,
    },
  });
}
