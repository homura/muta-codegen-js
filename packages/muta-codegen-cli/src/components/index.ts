export { generateEventCode } from './event';
export { generateServiceBindingCode } from './service';
export { generatePackageJSONCode } from './packageJSON';
