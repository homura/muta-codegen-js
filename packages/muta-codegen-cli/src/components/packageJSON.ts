interface PackageJSONOptions {
  name: string;
  mutaSDKVersion: string;
}

export function generatePackageJSONCode(
  options: Partial<PackageJSONOptions>,
): string {
  return `{
  "name": "${options.name}",
  "main": "lib/index.js",
  "scripts": {
    "build": "tsc --outDir lib src/**/*.ts"
  },
  "dependencies": {
    "@mutajs/types": "${options.mutaSDKVersion}",
    "@mutajs/service": "${options.mutaSDKVersion}"
  },
  "devDependencies": {
    "typescript": "^3.8.0"
  }
} `;
}
