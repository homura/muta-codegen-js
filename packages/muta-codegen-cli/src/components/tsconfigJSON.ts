export function generateTsconfigJSONCode(): string {
  return `{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "outDir": "lib"
  },
  "exclude": [
    "node_modules"
  ],
  "include": ["./src"]
}`;
}
