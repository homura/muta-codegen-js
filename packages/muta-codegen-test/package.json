{
  "name": "muta-codegen-test",
  "version": "0.0.1",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "private": true,
  "files": [
    "lib"
  ],
  "scripts": {
    "clean": "rimraf lib",
    "test": "jest",
    "test:watch": "jest --watch",
    "coverage": "jest --coverage",
    "lint": "eslint",
    "lint:fix": "eslint --fix",
    "precommit": "lint-staged"
  },
  "dependencies": {
  }
}
