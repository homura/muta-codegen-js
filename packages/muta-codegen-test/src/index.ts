export { AssetServiceSchema } from './mock/AssetService.graphql';
export { AssetEventSchema } from './mock/AssetEvent.graphql';
