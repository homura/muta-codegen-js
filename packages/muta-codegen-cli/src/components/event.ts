import * as typescriptPlugin from '@graphql-codegen/typescript';
import { event as eventPlugin } from 'graphql-codegen-plugin-mutasdk';
import { codegen } from './basic';

export async function generateEventCode(eventSource: string): Promise<string> {
  return codegen({
    pluginsMap: {
      typescript: typescriptPlugin,
      event: eventPlugin,
    },
    plugins: [{ event: {} }, { typescript: {} }],
    schemaSource: eventSource,
  });
}
