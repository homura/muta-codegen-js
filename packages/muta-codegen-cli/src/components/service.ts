import * as typescriptPlugin from '@graphql-codegen/typescript';
import { service as servicePlugin } from 'graphql-codegen-plugin-mutasdk';
import { codegen } from './basic';

export async function generateServiceBindingCode(
  serviceName: string,
  schemaSource: string,
): Promise<string> {
  return codegen({
    pluginsMap: {
      typescript: typescriptPlugin,
      service: servicePlugin,
    },
    plugins: [{ service: { service: serviceName } }, { typescript: {} }],
    schemaSource,
  });
}
