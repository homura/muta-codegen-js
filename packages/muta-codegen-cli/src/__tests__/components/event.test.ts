import { AssetEventSchema } from 'muta-codegen-test';
import { generateEventCode } from '../../components';

test('simple event should codegen correct', async () => {
  const source = `
type CreateAssetEvent {
    asset_id: String!
}
`;
  const schema = await generateEventCode(source);
  expect(schema).toMatchSnapshot();
});

test('empty when no event schema', async () => {
  const schema = await generateEventCode('');
  expect(schema).toEqual('');
});

test('asset event schema', async () => {
  expect(await generateEventCode(AssetEventSchema)).toMatchSnapshot();
});
