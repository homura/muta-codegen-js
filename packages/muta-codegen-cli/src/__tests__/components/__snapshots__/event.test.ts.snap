// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asset event schema 1`] = `
"// generated by muta-codegen, please do not edit manually
/* eslint-disable */
import { u64, Hash, Address, Uint64 , Bytes, u32, Vec } from \\"@mutajs/types\\";

export type Maybe<T> = T | null;


export function isEventOfAsset(event: any): event is Asset {
  return event?.topic === 'Asset';
}

export function isEventOfTransferEvent(event: any): event is TransferEvent {
  return event?.topic === 'TransferEvent';
}

export function isEventOfApproveEvent(event: any): event is ApproveEvent {
  return event?.topic === 'ApproveEvent';
}

export function isEventOfTransferFromEvent(event: any): event is TransferFromEvent {
  return event?.topic === 'TransferFromEvent';
}

type _EventName = 'Asset' |
  'TransferEvent' |
  'ApproveEvent' |
  'TransferFromEvent';

interface _IsEventOf {
  (name: 'Asset', event: any): event is Asset;
  (name: 'TransferEvent', event: any): event is TransferEvent;
  (name: 'ApproveEvent', event: any): event is ApproveEvent;
  (name: 'TransferFromEvent', event: any): event is TransferFromEvent;
}

export const isEventOf: _IsEventOf =
  ((name: _EventName, event: any) => event?.topic === name) as _IsEventOf;
  
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Address: Address;
  Hash: Hash;
  Uint64: Uint64;
};


export type ApproveEvent = {
  asset_id: Scalars['Hash'];
  grantor: Scalars['Address'];
  grantee: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type Asset = {
  id: Scalars['Hash'];
  name: Scalars['String'];
  symbol: Scalars['String'];
  supply: Scalars['Uint64'];
  issuer: Scalars['Address'];
};


export type TransferEvent = {
  asset_id: Scalars['Hash'];
  from: Scalars['Address'];
  to: Scalars['Address'];
  value: Scalars['Uint64'];
};

export type TransferFromEvent = {
  asset_id: Scalars['Hash'];
  caller: Scalars['Address'];
  sender: Scalars['Address'];
  recipient: Scalars['Address'];
  value: Scalars['Uint64'];
};


export type Event = Asset | TransferEvent | ApproveEvent | TransferFromEvent;
"
`;

exports[`simple event should codegen correct 1`] = `
"// generated by muta-codegen, please do not edit manually
/* eslint-disable */
import { u64, Hash, Address, Uint64 , Bytes, u32, Vec } from \\"@mutajs/types\\";

export type Maybe<T> = T | null;

/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CreateAssetEvent = {
  asset_id: Scalars['String'];
};
"
`;
